Test+misc

leamas opened the pull request:
This is  a pretty large PR:
```
$ git diff --stat upstream/master
24 files changed, 1424 insertions(+), 547 deletions(-)
```
I'm pretty confident that the changes are OK. However, I will not be around to support this over the weekend which is filled with social activities.

The main points:
  - Added framework for off-line n2k tests (too many hours spent on this).
  - Cleaned up comm_drv_registry. Notably, the singleton implementation was soppy, leading to the errors when using auto. This hole is now plugged.
  - Major clean up of comm_drv_n2k_socketcan. Most of these changes should be applicable to other drivers as well.

The driver update:
  - Overall clean up of headers, variable names, etc.
  - Use wxThread -> std::thread. Basically a test issue, but IMHO general C++ mechanisms should be preferred over wxWidgets' when applicable. C++ has a somewhat larger community.
  - Clean up driver deactivation logic, make sure a deleted driver does not linger in the registry.
  - Fix a nasty bug where _make_shared_from_this()_ was called in a destructor, basically undefined behaviour. 
  - Re-factor the generic atomic_queue to a separate header.
  - Post the received message directly in the worker thread, omitting the round trip to the parent driver. This is possible since observable is thread safe.
  - Replace strange _select(2)_ with  _read(2)_. No need to multiplex io here, blocking read is much simpler.
  - Rewrite the error checks from select/read to something more sane.
  - Remove quite a large amount of dead code, useless evil #ifdefs etc.
  - Re-factor the too large worker thread function.
  
... and what not.

bdbcat commented:
Looks pretty good.  I appreciate the cleanup and simplification.
One question:  How can we prove that observable structure is truly thread safe?  Simple code read/analysis?
Could we contrive a test case to stress this hypothesis?
I'll wait until you are a bit more free to merge this.

bdbcat commented:
If you don't mind, I'll cherry pick https://github.com/OpenCPN/OpenCPN/pull/2803/commits/767894e1f9c82244da1801bc99512a167961489c now.
Thanks
Dave

leamas commented:
>One question: How can we prove that observable structure is truly thread safe? Simple code read/analysis?
Could we contrive a test case to stress this hypothesis?

Good question. Let's try both.

I will try to create a testcase, it shouldn't be that hard to spin off some  stress testing threads. 

As for reading, I have done what I can. It's all about observable.cpp. My understanding is that there are two critical regions:

1.  The instances map, static local in GetInstance().
2.  Each instance has a m_listeners map, also a critical region. This is accessed when doing Listen(), Unlisten() and Notify().

It would of course be great if you could have a look as well, four eyes sees more than two.

One a sidenote, this is a readers-writers case which could be optimized using read and write locks. However, I don't think it's worth it, there should not be much congestions on these locks in real life.

bdbcat commented:
Thinking about thread safety overnight.
I'm pretty sure this patch is going to be a problem.  Notifying a listener from a thread other than the main wxWidgets thread will be trouble.  Consider this scenario:
1.  Assume N2K worker thread notifies listeners directly, as this patch does.  One listener is comm_bridge.  So, IIUC, the comm bridge lambda will be executed in the context of the n2k worker thread.
2. There is another worker thread for, say, N0183 serial.  He also might notify the listener in comm_bridge, at any asynchronous time.
3. So, we have two threads banging on comm_bridge.  Theoretically could be made to work with some mutex's in comm_bridge.  But why?, since...
4. The same problem, or worse, will happen for example with AIS decoder.  Again, we could fix AIS decoder with mutex or critical section.
5. Finally, we will need some main thread signaling somewhere, since we cannot do anything GUI related in any thread other than the main wxWidgets app thread.

Toughts?

leamas commented:
> So, IIUC, the comm bridge lambda will be executed in the context of the n2k worker thread.

No. The worker thread does wxQueueEvent, and only that. The actual processing of the event, in this case the comm_bridge lambda, happens when processing the event queued by   the worker thread. This processing happens in the main thread. 

And since the processing of the queue in the main thread is sequential, there are no collisions.

Or?

bdbcat commented:
OK, I'm satisfied.  I have not read the notify() code closely.  So many LOC.

leamas commented:
Trying to find some time to create a test case. Seems to take more time than anticipated (surprise!)

leamas commented:
I have found two more threading issues in the socketcan driver, perhaps corner cases but still... In one case  raw pointers/new/delete once again causes inconsistencies. I'm on it, but is also renovating my kitchen, time is an issue.

OTOH, it is so far simple to keep the branch rebased.

leamas commented:
I have added more work to this. More highlights:
  - The idea to dynamically allocate the thread is strange and dangerous, opens for nasty bugs where a running thread uses deleted memory as code. I have made the thread code a normal class variable, always allocated and with knops to start and stop the thread. 
  - There was some issues with data shared between main and worker thread. wxString assignments are not thread safe, and even if an integer probably is it should be using `std::atomic` to play it safe. Fixed
  - In the process, almost everything is moved to the cpp file. The header is really small which is a Good Thing.
  - I have refactored the driver code once again, moving the FastMessages stuff to a separate class.
  - All raw pointers `new` calls etc. are gone. The hoem-grown FastMessages memory management is replaced with regular vectors.
  - Time and garbage collection is cleaned up, based on std::chrono rather than wxWidgets (I tend to trust C++ standard libs more than the wx ones).

I'm done for now, please feel free to review, comment, merge or reject...

leamas commented:
hm... some homework with the MacOS builds and tests. I'll have a look tomorrow.

bdbcat commented:
Looks ...intense...
I'm going to hold this until after Alpha release (tonight).
Please don't be tempted to work on comm_nk2_serial.cpp just yet.  I am using this as testbed for writing on n2k bus, and I don't need a big rebase now.

leamas commented:
To merge this after the alpha seems reasonable.  It should be easy to keep it rebased until then.

I'm definitely not tempted to apply any changes from this anywhere else until it's merged, one step at a time. After all, this is quite an update. Still, I have confidence in it:  a simply better structure, more compile time checks, descent unit test coverage,  fixed bugs. The kind of things which happens on the second iteration.

leamas commented:
Rebased to current master, some more cleanup and refactoring.

leamas commented:
Rebased to current master

bdbcat commented:
I am ready for this, if you are.
Dave

leamas commented:
"ready"... there is always more to do. But I do think socketcan now represents a better driver structure which over time can be applied also to other drivers.

This is a lot of commits, but they  are all basically about the socketcan driver and the tests.  The exception is  b4cb5d71a719a153 which makes some basic clean up in the serial driver and also uses std::thread instead of wxThread. I'm not particularly nervous about it, but it is certainly possible to revert this in case of problems. 

The last commit is also affecting several drivers, but is more theoretical since SetListener() isn't (yet) actually used.

TL;DR yes, I'm ready
